[Cocoa] Use new screen/window picker API when available

eric-carlson opened the pull request:
#### fe99ba16036bd19b8cc1cff96d394c5dd5b1d275
<pre>
[Cocoa] Use new screen/window picker API when available
<a href="https://bugs.webkit.org/show_bug.cgi?id=251893">https://bugs.webkit.org/show_bug.cgi?id=251893</a>
rdar://101206117

Reviewed by Jer Noble.

* Source/WebCore/Headers.cmake: Add DisplayCapturePromptType.h.

* Source/WebCore/WebCore.xcodeproj/project.pbxproj: Ditto.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml: SC_CONTENT_SHARING_SESSION -&gt; SCREEN_CAPTURE_KIT.
Add new runtime preference.

* Source/WTF/wtf/PlatformHave.h: Remove HAVE_SC_CONTENT_SHARING_SESSION.

* Source/WebCore/PAL/pal/mac/ScreenCaptureKitSoftLink.h:SCStreamFrameInfoStatus
is always available when ScreenCaptureKit is available.
* Source/WebCore/PAL/pal/mac/ScreenCaptureKitSoftLink.mm:
* Source/WebCore/PAL/pal/spi/mac/ScreenCaptureKitSPI.h:

* Source/WebKit/Shared/Cocoa/WebPreferencesDefaultValuesCocoa.mm:
(WebKit::defaultUseSCContentSharingPicker): Initialize to feature flag value.
* Source/WebKit/Shared/WebPreferencesDefaultValues.h:

* Source/WebCore/platform/audio/PlatformMediaSessionManager.cpp:
(WebCore::PlatformMediaSessionManager::setUseSCContentSharingPicker): New preference access.
(WebCore::PlatformMediaSessionManager::useSCContentSharingPicker):
* Source/WebCore/platform/audio/PlatformMediaSessionManager.h:

* Source/WebCore/platform/mediastream/mac/ScreenCaptureKitCaptureSource.h:
* Source/WebCore/platform/mediastream/mac/ScreenCaptureKitCaptureSource.mm:
Rewrite to not be a direct `SCContentSharingSession` delegate and to not create `SCContentStream`
directly, but to do both through the ScreenCaptureKitSharingSessionManager.h.

(-[WebCoreScreenCaptureKitHelper stream:didStopWithError:]):
(-[WebCoreScreenCaptureKitHelper stream:didOutputSampleBuffer:ofType:]):
(WebCore::ScreenCaptureKitCaptureSource::~ScreenCaptureKitCaptureSource):
(WebCore::ScreenCaptureKitCaptureSource::stop):
(WebCore::ScreenCaptureKitCaptureSource::sessionFailedWithError):
(WebCore::ScreenCaptureKitCaptureSource::sessionFilterDidChange):
(WebCore::ScreenCaptureKitCaptureSource::sessionStreamDidEnd):
(WebCore::ScreenCaptureKitCaptureSource::startContentStream):
(WebCore::ScreenCaptureKitCaptureSource::updateStreamConfiguration):
(WebCore::ScreenCaptureKitCaptureSource::commitConfiguration):
(WebCore::ScreenCaptureKitCaptureSource::streamDidOutputVideoSampleBuffer):
(-[WebCoreScreenCaptureKitHelper sessionDidEnd:]): Deleted.
(-[WebCoreScreenCaptureKitHelper sessionDidChangeContent:]): Deleted.
(-[WebCoreScreenCaptureKitHelper pickerCanceledForSession:]): Deleted.
(WebCore::ScreenCaptureKitCaptureSource::streamFailedWithError): Deleted.
(WebCore::ScreenCaptureKitCaptureSource::sessionDidChangeContent): Deleted.
(WebCore::ScreenCaptureKitCaptureSource::sessionDidEnd): Deleted.
(WebCore::findSharableDevice): Deleted.
(WebCore::ScreenCaptureKitCaptureSource::findShareableContent): Deleted.
(WebCore::ScreenCaptureKitCaptureSource::frameAvailableHandler): Deleted.
(WebCore::ScreenCaptureKitCaptureSource::streamDidOutputSampleBuffer): Deleted.
* Source/WebCore/platform/mediastream/mac/ScreenCaptureKitSharingSessionManager.h:
(WebCore::ScreenCaptureSessionSource::stream const):
(WebCore::ScreenCaptureSessionSource::contentFilter const):
(WebCore::ScreenCaptureSessionSource::sharingSession const):
(WebCore::ScreenCaptureSessionSource::observer const):

* Source/WebCore/platform/mediastream/mac/ScreenCaptureKitSharingSessionManager.mm:
Don&apos;t pass ownership of `SCContentSharingSession` to ScreenCaptureSessionSource, instead
pass necessary information through ScreenCaptureSessionSource interface to source.

(-[WebDisplayMediaPromptHelper initWithCallback:]):
(-[WebDisplayMediaPromptHelper sessionDidEnd:]):
(-[WebDisplayMediaPromptHelper sessionDidChangeContent:]):
(-[WebDisplayMediaPromptHelper pickerCanceledForSession:]):
(-[WebDisplayMediaPromptHelper contentSharingPicker:didSelectFilter:forStream:]):
(-[WebDisplayMediaPromptHelper contentSharingPickerDidCancel:forStream:]):
(-[WebDisplayMediaPromptHelper contentSharingPickerStartDidFailWithError:]):
(-[WebDisplayMediaPromptHelper contentSharingPicker:didRequestNewStreamWithFilter:]):
(-[WebDisplayMediaPromptHelper contentSharingPicker:didRequestUpdateWithFilter:forStream:]):
(WebCore::ScreenCaptureKitSharingSessionManager::isAvailable):
(WebCore::ScreenCaptureKitSharingSessionManager::useSCContentSharingPicker):
(WebCore::ScreenCaptureKitSharingSessionManager::~ScreenCaptureKitSharingSessionManager):
(WebCore::ScreenCaptureKitSharingSessionManager::cancelPicking):
(WebCore::ScreenCaptureKitSharingSessionManager::completeDeviceSelection):
(WebCore::ScreenCaptureKitSharingSessionManager::findActiveSource):
(WebCore::ScreenCaptureKitSharingSessionManager::sharingSessionDidEnd):
(WebCore::ScreenCaptureKitSharingSessionManager::sharingSessionDidChangeContent):
(WebCore::ScreenCaptureKitSharingSessionManager::contentSharingPickerSelectedFilterForStream):
(WebCore::ScreenCaptureKitSharingSessionManager::contentSharingPickerUpdatedFilterForStream):
(WebCore::ScreenCaptureKitSharingSessionManager::promptForGetDisplayMedia):
(WebCore::ScreenCaptureKitSharingSessionManager::promptWithSCContentSharingSession):
(WebCore::ScreenCaptureKitSharingSessionManager::promptWithSCContentSharingPicker):
(WebCore::ScreenCaptureKitSharingSessionManager::contentSharingPickerFailedWithError):
(WebCore::operator==):
(WebCore::ScreenCaptureKitSharingSessionManager::createSessionSourceForDevice):
(WebCore::ScreenCaptureKitSharingSessionManager::cleanupSessionSource):
(WebCore::ScreenCaptureSessionSource::create):
(WebCore::ScreenCaptureSessionSource::ScreenCaptureSessionSource):
(WebCore::ScreenCaptureSessionSource::~ScreenCaptureSessionSource):
(WebCore::ScreenCaptureSessionSource::operator== const):
(WebCore::screenCaptureKitPickerFeatureEnabled): Deleted.
(WebCore::ScreenCaptureKitSharingSessionManager::pickerCanceledForSession): Deleted.
(WebCore::ScreenCaptureKitSharingSessionManager::sessionDidEnd): Deleted.
(WebCore::ScreenCaptureKitSharingSessionManager::sessionDidChangeContent): Deleted.
(WebCore::ScreenCaptureKitSharingSessionManager::showWindowPicker): Deleted.
(WebCore::ScreenCaptureKitSharingSessionManager::showScreenPicker): Deleted.
(WebCore::ScreenCaptureKitSharingSessionManager::takeSharingSessionForFilter): Deleted.

* Source/WebKit/GPUProcess/GPUProcess.cpp:
(WebKit::GPUProcess::promptForGetDisplayMedia): Combine showWindowPicker and showScreenPicker.
(WebKit::GPUProcess::showWindowPicker): Deleted.
(WebKit::GPUProcess::showScreenPicker): Deleted.
* Source/WebKit/GPUProcess/GPUProcess.h:
* Source/WebKit/GPUProcess/GPUProcess.messages.in:
* Source/WebKit/GPUProcess/mac/com.apple.WebKit.GPUProcess.sb.in:

* Source/WebKit/UIProcess/Cocoa/UIDelegate.mm:
(WebKit::UIDelegate::UIClient::promptForDisplayCapturePermission): Cleanup and simplify.
(WebKit::UIDelegate::UIClient::decidePolicyForUserMediaPermissionRequest):

* Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm: SC_CONTENT_SHARING_SESSION -&gt; SCREEN_CAPTURE_KIT

* Source/WebKit/UIProcess/GPU/GPUProcessProxy.cpp:
(WebKit::GPUProcessProxy::updateSandboxAccess): Ditto.
(WebKit::GPUProcessProxy::promptForGetDisplayMedia):
(WebKit::GPUProcessProxy::showWindowPicker): Deleted.
(WebKit::GPUProcessProxy::showScreenPicker): Deleted.
* Source/WebKit/UIProcess/GPU/GPUProcessProxy.h:

* Source/WebKit/UIProcess/UserMediaPermissionRequestProxy.cpp:
(WebKit::UserMediaPermissionRequestProxy::promptForGetDisplayMedia): canPromptForGetDisplayMedia -&gt;
canRequestDisplayCapturePermission.
(WebKit::UserMediaPermissionRequestProxy::doDefaultAction):
(WebKit::UserMediaPermissionRequestProxy::canRequestDisplayCapturePermission): Rename from
canPromptForGetDisplayMedia.
(WebKit::UserMediaPermissionRequestProxy::canPromptForGetDisplayMedia): Deleted.
* Source/WebKit/UIProcess/UserMediaPermissionRequestProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:

* Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivateForTesting.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewTesting.mm:
(-[WKWebView _setSystemCanPromptForGetDisplayMediaForTesting:]): Simulate when system can
and can not prompt for type of display capture.

* Source/WebKit/UIProcess/mac/DisplayCaptureSessionManager.h:
* Source/WebKit/UIProcess/mac/DisplayCaptureSessionManager.mm:
(WebKit::DisplayCaptureSessionManager::showWindowPicker):
(WebKit::DisplayCaptureSessionManager::showScreenPicker):
(WebKit::DisplayCaptureSessionManager::canRequestDisplayCapturePermission):
(WebKit::toScreenCaptureKitPromptType):
(WebKit::DisplayCaptureSessionManager::promptForGetDisplayMedia):

* Source/WebKit/UIProcess/mac/UserMediaPermissionRequestProxyMac.h:
* Source/WebKit/UIProcess/mac/UserMediaPermissionRequestProxyMac.mm:
(WebKit::UserMediaPermissionRequestProxyMac::promptForGetDisplayMedia):
(WebKit::UserMediaPermissionRequestProxyMac::canRequestDisplayCapturePermission):
(WebKit::UserMediaPermissionRequestProxyMac::canPromptForGetDisplayMedia): Deleted.
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::updatePreferences):

* Tools/TestWebKitAPI/Tests/WebKitCocoa/GetDisplayMedia.mm: Test when system can and can
not prompt for type of capture.
(-[GetDisplayMediaUIDelegate _webView:requestDisplayCapturePermissionForOrigin:initiatedByFrame:withSystemAudio:decisionHandler:]):
(TestWebKitAPI::GetDisplayMediaTest::promptForCapture):
(TestWebKitAPI::TEST_F):
(-[GetDisplayMediaUIDelegate _webView:requestUserMediaAuthorizationForDevices:url:mainFrameURL:decisionHandler:]): Deleted.

Canonical link: <a href="https://commits.webkit.org/261180@main">https://commits.webkit.org/261180@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e37c1d967a18558844bbd347ac2f07f7feb6946f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/110809 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/19895 "Built successfully") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/43/builds/43354 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/87/builds/2157 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/119622 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/114772 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/21293 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/10997 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/87/builds/2157 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/116559 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/15869 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/43/builds/43354 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/103170 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/97829 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/43/builds/43354 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/44244 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/1/builds/99454 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/81/builds/12487 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/43/builds/43354 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/86125 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/10595 "Built successfully and passed tests") | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/82/builds/13053 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/9007 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/100541 "Built successfully") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/18436 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/43/builds/43354 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/31356 "Passed tests") | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/74/builds/7752 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/79/builds/15007 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/108571 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/4224 "Built successfully and passed tests") | | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/26763 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebCore/platform/mediastream/mac/ScreenCaptureKitCaptureSource.mm:
```diff
@@ -500,52 +381,27 @@ static void findSharableDevice(RetainPtr<SCShareableContent>&& shareableContent,
+void ScreenCaptureKitCaptureSource::streamDidOutputVideoSampleBuffer(RetainPtr<CMSampleBufferRef> sampleBuffer)
 {
     ASSERT(isMainThread());
+    ASSERT(sampleBuffer);
```
We should avoid going through main thread at some point.

eric-carlson replied:
Yes, lets do that in a followup.

youennf commented on Source/WebCore/platform/mediastream/mac/ScreenCaptureKitSharingSessionManager.mm:
```diff
@@ -268,56 +383,188 @@ static bool screenCaptureKitPickerFeatureEnabled()
+    bool showingPicker = useSCContentSharingPicker() ? promptWithSCContentSharingPicker(promptType) : promptWithSCContentSharingSession(promptType);
+    if (!showingPicker) {
+        m_completionHandler(std::nullopt);
+        m_completionHandler = { };
```
m_completionHandler is nullifying itself when being called so  this line is probably not needed.

youennf commented on Source/WebCore/PAL/pal/mac/ScreenCaptureKitSoftLink.h:
```diff
@@ -37,11 +37,9 @@ SOFT_LINK_CLASS_FOR_HEADER_WITH_AVAILABILITY(PAL, SCContentFilter, API_AVAILABLE
 SOFT_LINK_CLASS_FOR_HEADER_WITH_AVAILABILITY(PAL, SCStreamConfiguration, API_AVAILABLE(macos(12.3)))
 SOFT_LINK_CLASS_FOR_HEADER_WITH_AVAILABILITY(PAL, SCStream, API_AVAILABLE(macos(12.3)))
 SOFT_LINK_CLASS_FOR_HEADER(PAL, SCContentSharingSession)
+SOFT_LINK_CLASS_FOR_HEADER(PAL, SCContentSharingPicker)
```
Probably not a big deal but should it be covered by HAVE_SC_CONTENT_SHARING_PICKER?

youennf commented on Source/WebCore/platform/mediastream/DisplayCapturePromptType.h:
```diff
@@ -0,0 +1,49 @@
+/*
+ * Copyright (C) 2021-2022 Apple Inc. All rights reserved.
```
2023

youennf commented on Source/WebCore/platform/mediastream/mac/ScreenCaptureKitCaptureSource.h:
```diff
@@ -84,28 +82,28 @@ class ScreenCaptureKitCaptureSource final
 
+    // ScreenCaptureKitSharingSessionManager::Observer
+    void sessionFilterDidChange(RetainPtr<SCContentFilter>) final;
+    void sessionStreamDidEnd(RetainPtr<SCStream>) final;
```
Could be &&

eric-carlson replied:
I changed them to take pointers

youennf commented on Source/WebCore/platform/mediastream/mac/ScreenCaptureKitCaptureSource.mm:
```diff
@@ -108,41 +81,47 @@ - (void)stream:(SCStream *)stream didStopWithError:(NSError *)error
-        if (!strongSelf->_callback)
+    ASSERT(type == SCStreamOutputTypeScreen);
+
+    if (!sampleBuffer)
```
In which case can sampleBuffer be null?

youennf commented on Source/WebCore/platform/mediastream/mac/ScreenCaptureKitSharingSessionManager.mm:
```diff
@@ -242,82 +279,298 @@ static bool screenCaptureKitPickerFeatureEnabled()
+    , m_contentFilter(WTFMove(filter))
+    , m_sharingSession(WTFMove(sharingSession))
+    , m_observer(WTFMove(observer))
+    , m_cleanupFunction(WTFMove(cleanupFunction))
```
It should probably be a CompletionHandler<> instead of a Function.

eric-carlson replied:
Fixed

youennf commented on Source/WebCore/platform/mediastream/mac/ScreenCaptureKitSharingSessionManager.mm:
```diff
@@ -153,87 +219,58 @@ static bool screenCaptureKitPickerFeatureEnabled()
+    if (!contentFilter || !m_completionHandler) {
+        if (m_completionHandler)
+            m_completionHandler(std::nullopt);
+        m_completionHandler = { };
```
Line probably not needed

youennf commented on Source/WebCore/platform/mediastream/mac/ScreenCaptureKitSharingSessionManager.mm:
```diff
@@ -242,82 +279,298 @@ static bool screenCaptureKitPickerFeatureEnabled()
 
-    RELEASE_LOG(WebRTC, "ScreenCaptureKitSharingSessionManager::takeSharingSessionForFilter");
+    if (error)
+        RELEASE_LOG_ERROR(WebRTC, "ScreenCaptureKitSharingSessionManager::contentSharingPickerFailedWithError content sharing picker failed with error '%s'", [[error localizedDescription] UTF8String]);
```
RELEASE_LOG_ERROR_IF

youennf commented on Source/WebKit/UIProcess/GPU/GPUProcessProxy.cpp:
```diff
@@ -362,15 +375,10 @@ void GPUProcessProxy::triggerMockMicrophoneConfigurationChange()
+void GPUProcessProxy::promptForGetDisplayMedia(WebCore::DisplayCapturePromptType type, CompletionHandler<void(std::optional<WebCore::CaptureDevice>)>&& completionHandler)
 {
-    sendWithAsyncReply(Messages::GPUProcess::ShowScreenPicker { }, WTFMove(completionHandler));
+    sendWithAsyncReply(Messages::GPUProcess::PromptForGetDisplayMedia { type }, WTFMove(completionHandler));
```
Could we pass m_useSCContentSharingPicker here instead of the separate IPC?

cdumez commented on Source/WebCore/platform/audio/PlatformMediaSessionManager.h:
```diff
@@ -238,6 +240,9 @@ class PlatformMediaSessionManager
     static bool m_alternateWebMPlayerEnabled;
 #endif
+#if HAVE(SC_CONTENT_SHARING_PICKER)
+    static bool m_useSCContentSharingPicker;
```
Not new to this patch but it is a little weird to use `m_` prefix for static members. I think we often use `s_` prefix for those.

eric-carlson replied:
OK, changed.

cdumez commented on Source/WebCore/platform/mediastream/DisplayCapturePromptType.h:
```diff
@@ -0,0 +1,49 @@
+
+namespace WTF {
+
+template<> struct EnumTraits<WebCore::DisplayCapturePromptType> {
```
This shouldn't be needed if you used the new serialization.in file.

eric-carlson replied:
Good idea, changed.

eric-carlson committed:
[Cocoa] Use new screen/window picker API when available
https://bugs.webkit.org/show_bug.cgi?id=251893
rdar://101206117

Reviewed by Jer Noble.

* Source/WebCore/Headers.cmake: Add DisplayCapturePromptType.h.

* Source/WebCore/WebCore.xcodeproj/project.pbxproj: Ditto.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml: SC_CONTENT_SHARING_SESSION -> SCREEN_CAPTURE_KIT.
Add new runtime preference.

* Source/WTF/wtf/PlatformHave.h: Remove HAVE_SC_CONTENT_SHARING_SESSION.

* Source/WebCore/PAL/pal/mac/ScreenCaptureKitSoftLink.h:SCStreamFrameInfoStatus
is always available when ScreenCaptureKit is available.
* Source/WebCore/PAL/pal/mac/ScreenCaptureKitSoftLink.mm:
* Source/WebCore/PAL/pal/spi/mac/ScreenCaptureKitSPI.h:

* Source/WebKit/Shared/Cocoa/WebPreferencesDefaultValuesCocoa.mm:
(WebKit::defaultUseSCContentSharingPicker): Initialize to feature flag value.
* Source/WebKit/Shared/WebPreferencesDefaultValues.h:

* Source/WebCore/platform/audio/PlatformMediaSessionManager.cpp:
(WebCore::PlatformMediaSessionManager::setUseSCContentSharingPicker): New preference access.
(WebCore::PlatformMediaSessionManager::useSCContentSharingPicker):
* Source/WebCore/platform/audio/PlatformMediaSessionManager.h:

* Source/WebCore/platform/mediastream/mac/ScreenCaptureKitCaptureSource.h:
* Source/WebCore/platform/mediastream/mac/ScreenCaptureKitCaptureSource.mm:
Rewrite to not be a direct `SCContentSharingSession` delegate and to not create `SCContentStream`
directly, but to do both through the ScreenCaptureKitSharingSessionManager.h.

(-[WebCoreScreenCaptureKitHelper stream:didStopWithError:]):
(-[WebCoreScreenCaptureKitHelper stream:didOutputSampleBuffer:ofType:]):
(WebCore::ScreenCaptureKitCaptureSource::~ScreenCaptureKitCaptureSource):
(WebCore::ScreenCaptureKitCaptureSource::stop):
(WebCore::ScreenCaptureKitCaptureSource::sessionFailedWithError):
(WebCore::ScreenCaptureKitCaptureSource::sessionFilterDidChange):
(WebCore::ScreenCaptureKitCaptureSource::sessionStreamDidEnd):
(WebCore::ScreenCaptureKitCaptureSource::startContentStream):
(WebCore::ScreenCaptureKitCaptureSource::updateStreamConfiguration):
(WebCore::ScreenCaptureKitCaptureSource::commitConfiguration):
(WebCore::ScreenCaptureKitCaptureSource::streamDidOutputVideoSampleBuffer):
(-[WebCoreScreenCaptureKitHelper sessionDidEnd:]): Deleted.
(-[WebCoreScreenCaptureKitHelper sessionDidChangeContent:]): Deleted.
(-[WebCoreScreenCaptureKitHelper pickerCanceledForSession:]): Deleted.
(WebCore::ScreenCaptureKitCaptureSource::streamFailedWithError): Deleted.
(WebCore::ScreenCaptureKitCaptureSource::sessionDidChangeContent): Deleted.
(WebCore::ScreenCaptureKitCaptureSource::sessionDidEnd): Deleted.
(WebCore::findSharableDevice): Deleted.
(WebCore::ScreenCaptureKitCaptureSource::findShareableContent): Deleted.
(WebCore::ScreenCaptureKitCaptureSource::frameAvailableHandler): Deleted.
(WebCore::ScreenCaptureKitCaptureSource::streamDidOutputSampleBuffer): Deleted.
* Source/WebCore/platform/mediastream/mac/ScreenCaptureKitSharingSessionManager.h:
(WebCore::ScreenCaptureSessionSource::stream const):
(WebCore::ScreenCaptureSessionSource::contentFilter const):
(WebCore::ScreenCaptureSessionSource::sharingSession const):
(WebCore::ScreenCaptureSessionSource::observer const):

* Source/WebCore/platform/mediastream/mac/ScreenCaptureKitSharingSessionManager.mm:
Don't pass ownership of `SCContentSharingSession` to ScreenCaptureSessionSource, instead
pass necessary information through ScreenCaptureSessionSource interface to source.

(-[WebDisplayMediaPromptHelper initWithCallback:]):
(-[WebDisplayMediaPromptHelper sessionDidEnd:]):
(-[WebDisplayMediaPromptHelper sessionDidChangeContent:]):
(-[WebDisplayMediaPromptHelper pickerCanceledForSession:]):
(-[WebDisplayMediaPromptHelper contentSharingPicker:didSelectFilter:forStream:]):
(-[WebDisplayMediaPromptHelper contentSharingPickerDidCancel:forStream:]):
(-[WebDisplayMediaPromptHelper contentSharingPickerStartDidFailWithError:]):
(-[WebDisplayMediaPromptHelper contentSharingPicker:didRequestNewStreamWithFilter:]):
(-[WebDisplayMediaPromptHelper contentSharingPicker:didRequestUpdateWithFilter:forStream:]):
(WebCore::ScreenCaptureKitSharingSessionManager::isAvailable):
(WebCore::ScreenCaptureKitSharingSessionManager::useSCContentSharingPicker):
(WebCore::ScreenCaptureKitSharingSessionManager::~ScreenCaptureKitSharingSessionManager):
(WebCore::ScreenCaptureKitSharingSessionManager::cancelPicking):
(WebCore::ScreenCaptureKitSharingSessionManager::completeDeviceSelection):
(WebCore::ScreenCaptureKitSharingSessionManager::findActiveSource):
(WebCore::ScreenCaptureKitSharingSessionManager::sharingSessionDidEnd):
(WebCore::ScreenCaptureKitSharingSessionManager::sharingSessionDidChangeContent):
(WebCore::ScreenCaptureKitSharingSessionManager::contentSharingPickerSelectedFilterForStream):
(WebCore::ScreenCaptureKitSharingSessionManager::contentSharingPickerUpdatedFilterForStream):
(WebCore::ScreenCaptureKitSharingSessionManager::promptForGetDisplayMedia):
(WebCore::ScreenCaptureKitSharingSessionManager::promptWithSCContentSharingSession):
(WebCore::ScreenCaptureKitSharingSessionManager::promptWithSCContentSharingPicker):
(WebCore::ScreenCaptureKitSharingSessionManager::contentSharingPickerFailedWithError):
(WebCore::operator==):
(WebCore::ScreenCaptureKitSharingSessionManager::createSessionSourceForDevice):
(WebCore::ScreenCaptureKitSharingSessionManager::cleanupSessionSource):
(WebCore::ScreenCaptureSessionSource::create):
(WebCore::ScreenCaptureSessionSource::ScreenCaptureSessionSource):
(WebCore::ScreenCaptureSessionSource::~ScreenCaptureSessionSource):
(WebCore::ScreenCaptureSessionSource::operator== const):
(WebCore::screenCaptureKitPickerFeatureEnabled): Deleted.
(WebCore::ScreenCaptureKitSharingSessionManager::pickerCanceledForSession): Deleted.
(WebCore::ScreenCaptureKitSharingSessionManager::sessionDidEnd): Deleted.
(WebCore::ScreenCaptureKitSharingSessionManager::sessionDidChangeContent): Deleted.
(WebCore::ScreenCaptureKitSharingSessionManager::showWindowPicker): Deleted.
(WebCore::ScreenCaptureKitSharingSessionManager::showScreenPicker): Deleted.
(WebCore::ScreenCaptureKitSharingSessionManager::takeSharingSessionForFilter): Deleted.

* Source/WebKit/GPUProcess/GPUProcess.cpp:
(WebKit::GPUProcess::promptForGetDisplayMedia): Combine showWindowPicker and showScreenPicker.
(WebKit::GPUProcess::showWindowPicker): Deleted.
(WebKit::GPUProcess::showScreenPicker): Deleted.
* Source/WebKit/GPUProcess/GPUProcess.h:
* Source/WebKit/GPUProcess/GPUProcess.messages.in:
* Source/WebKit/GPUProcess/mac/com.apple.WebKit.GPUProcess.sb.in:

* Source/WebKit/UIProcess/Cocoa/UIDelegate.mm:
(WebKit::UIDelegate::UIClient::promptForDisplayCapturePermission): Cleanup and simplify.
(WebKit::UIDelegate::UIClient::decidePolicyForUserMediaPermissionRequest):

* Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm: SC_CONTENT_SHARING_SESSION -> SCREEN_CAPTURE_KIT

* Source/WebKit/UIProcess/GPU/GPUProcessProxy.cpp:
(WebKit::GPUProcessProxy::updateSandboxAccess): Ditto.
(WebKit::GPUProcessProxy::promptForGetDisplayMedia):
(WebKit::GPUProcessProxy::showWindowPicker): Deleted.
(WebKit::GPUProcessProxy::showScreenPicker): Deleted.
* Source/WebKit/UIProcess/GPU/GPUProcessProxy.h:

* Source/WebKit/UIProcess/UserMediaPermissionRequestProxy.cpp:
(WebKit::UserMediaPermissionRequestProxy::promptForGetDisplayMedia): canPromptForGetDisplayMedia ->
canRequestDisplayCapturePermission.
(WebKit::UserMediaPermissionRequestProxy::doDefaultAction):
(WebKit::UserMediaPermissionRequestProxy::canRequestDisplayCapturePermission): Rename from
canPromptForGetDisplayMedia.
(WebKit::UserMediaPermissionRequestProxy::canPromptForGetDisplayMedia): Deleted.
* Source/WebKit/UIProcess/UserMediaPermissionRequestProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:

* Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivateForTesting.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewTesting.mm:
(-[WKWebView _setSystemCanPromptForGetDisplayMediaForTesting:]): Simulate when system can
and can not prompt for type of display capture.

* Source/WebKit/UIProcess/mac/DisplayCaptureSessionManager.h:
* Source/WebKit/UIProcess/mac/DisplayCaptureSessionManager.mm:
(WebKit::DisplayCaptureSessionManager::showWindowPicker):
(WebKit::DisplayCaptureSessionManager::showScreenPicker):
(WebKit::DisplayCaptureSessionManager::canRequestDisplayCapturePermission):
(WebKit::toScreenCaptureKitPromptType):
(WebKit::DisplayCaptureSessionManager::promptForGetDisplayMedia):

* Source/WebKit/UIProcess/mac/UserMediaPermissionRequestProxyMac.h:
* Source/WebKit/UIProcess/mac/UserMediaPermissionRequestProxyMac.mm:
(WebKit::UserMediaPermissionRequestProxyMac::promptForGetDisplayMedia):
(WebKit::UserMediaPermissionRequestProxyMac::canRequestDisplayCapturePermission):
(WebKit::UserMediaPermissionRequestProxyMac::canPromptForGetDisplayMedia): Deleted.
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::updatePreferences):

* Tools/TestWebKitAPI/Tests/WebKitCocoa/GetDisplayMedia.mm: Test when system can and can
not prompt for type of capture.
(-[GetDisplayMediaUIDelegate _webView:requestDisplayCapturePermissionForOrigin:initiatedByFrame:withSystemAudio:decisionHandler:]):
(TestWebKitAPI::GetDisplayMediaTest::promptForCapture):
(TestWebKitAPI::TEST_F):
(-[GetDisplayMediaUIDelegate _webView:requestUserMediaAuthorizationForDevices:url:mainFrameURL:decisionHandler:]): Deleted.

Canonical link: https://commits.webkit.org/261180@main
